src/chunk.c

```diff
@@ -4927,7 +4940,40 @@ ts_osm_chunk_range_is_invalid(int64 range_start, int64 range_end)
 int32
 ts_chunk_get_osm_slice_id(int32 chunk_id, int32 time_dim_id)
 {
-	Chunk *chunk = ts_chunk_get_by_id(chunk_id, true);
+	ScanKeyData scankey[1];
+	static const DisplayKeyData displaykey[1] = {
+		[0] = { .name = "id", .as_string = DatumGetInt32AsString },
+	};
+
+	/*
+	 * Perform an index scan on chunk id.
+	 */
+	ScanKeyInit(&scankey[0],
+				Anum_chunk_idx_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(chunk_id));
+
+	/*
+	 * Read the dimension slice id for this chunk without locking the tuple.
+	 * The reason for this is that if we take the FOR KEY SHARE lock on the tuple
+	 * that ts_hypercube_from_constraints otherwise takes, then two concurrent
+	 * updates of the OSM chunk range will deadlock: They are both granted the
+	 * FOR KEY SHARE lock and then request an UPDATE lock when trying to update
+	 * the tuple. However, they both need to wait on one another to acquire it,
+	 * since it conflicts with the FOR KEY SHARE lock already held by the other
+	 * session.
+	 * Hence, we skip taking the shared lock here and instead take a FOR UPDATE
+	 * lock on the dimension slice later.
+	 */
+	Chunk *chunk = chunk_scan_find(CHUNK_ID_INDEX,
+								   scankey,
+								   1,
+								   CurrentMemoryContext,
+								   true /*fail_if_not_found*/,
+								   displaykey,
+								   false /*lock_dimslice*/);
```

Comment:
Do we really need to "build" the whole chunk to just lock one dimension slice? I guess it should be enough to scan the `chunk_constraint` and `dimension_slice` tables.

Reply:
fixed now